File: server/webapp/WEB-INF/rails.new/app/presenters/api_v3/config/tasks/exec_task_representer.rb
##########################################################################
# Copyright 2016 ThoughtWorks, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
##########################################################################

module ApiV3
  module Config
    module Tasks
      class ExecTaskRepresenter < ApiV3::Config::Tasks::BaseTaskRepresenter
        alias_method :task, :represented
        ERROR_KEYS = {
          "buildFile"        => "build_file",
          "onCancelConfig"   => "on_cancel",
          "runIf"            => "run_if",
          "nantPath"         => "nant_path",
          "argListString"    => "arguments"
        }

        property :command
        collection :arguments, skip_nil: true, exec_context: :decorator
        property :args, skip_nil: true, exec_context: :decorator
        property :working_directory

        def arguments
          return nil if task.getArgList().isEmpty()
          task.getArgList().map { |arg| arg.getValue() }
        end

        def arguments=(value)
          task.setArgsList(value)
        end

        def args
          return nil if com.thoughtworks.go.util.StringUtil.isBlank(task.getArgs)
          task.getArgs
        end

        def args=(value)
          task.setArgs(value)
        end
      end
    end
  end
end

File: api/api-shared-v9/src/main/java/com/thoughtworks/go/apiv9/admin/shared/representers/stages/tasks/ExecTaskRepresenter.java
/*
 * Copyright 2020 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.thoughtworks.go.apiv9.admin.shared.representers.stages.tasks;

import com.thoughtworks.go.api.base.OutputWriter;
import com.thoughtworks.go.api.representers.JsonReader;
import com.thoughtworks.go.config.Argument;
import com.thoughtworks.go.config.ExecTask;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class ExecTaskRepresenter {
    public static void toJSON(OutputWriter jsonWriter, ExecTask execTask) {
        BaseTaskRepresenter.toJSON(jsonWriter, execTask);
        jsonWriter.add("command", execTask.command());
        if (execTask.getArgList().isEmpty()) {
            jsonWriter.addIfNotNull("args", execTask.getArgs());

        } else {
            List<String> arguments = execTask.getArgList().stream().map(Argument::getValue).collect(Collectors.toList());
            jsonWriter.addChildList("arguments", arguments);
        }
        jsonWriter.addIfNotNull("working_directory", execTask.workingDirectory());
    }

    public static ExecTask fromJSON(JsonReader jsonReader) {
        ExecTask execTask = new ExecTask();
        if (jsonReader == null) {
            return execTask;
        }
        BaseTaskRepresenter.fromJSON(jsonReader, execTask);
        jsonReader.readStringIfPresent("command", execTask::setCommand);
        jsonReader.readArrayIfPresent("arguments", arguments -> {
            ArrayList<String> argList = new ArrayList<>();
            arguments.forEach(argument -> {
                argList.add(argument.getAsString());
            });
            execTask.setArgsList(argList.toArray(new String[arguments.size()]));
        });
        jsonReader.readStringIfPresent("args", execTask::setArgs);
        jsonReader.readStringIfPresent("working_directory", execTask::setWorkingDirectory);

        return execTask;
    }
}
